api/src/test/java/org/motechproject/nms/api/osgi/KilkariControllerBundleIT.java

```diff
@@ -212,22 +220,145 @@ public void testCreateSubscriptionRequest() throws IOException, InterruptedExcep
+
+        Subscriber subscriber = subscriptionService.getSubscriber(callingNumber);
+        int numberOfSubsBefore = subscriber.getSubscriptions().size();
+
+        SimpleHttpClient.execHttpRequest(httpPost, HttpStatus.SC_OK, ADMIN_USERNAME, ADMIN_PASSWORD);
+
+        subscriber = subscriptionService.getSubscriber(callingNumber);
+        int numberOfSubsAfter = subscriber.getSubscriptions().size();
+
+        // No additional subscription should be created because subscriber already has an active subscription
+        // to this pack
+        assertEquals(numberOfSubsBefore, numberOfSubsAfter);
+    }
+
+    @Test
+    public void testCreateSubscriptionRequestDifferentPacks() throws IOException, InterruptedException {
+        setupData();
+        long callingNumber = 9999911122L;
+
+        HttpPost httpPost1 = createSubscriptionHttpPost(callingNumber, "pack1");
+
+        SimpleHttpClient.execHttpRequest(httpPost1, HttpStatus.SC_OK, ADMIN_USERNAME, ADMIN_PASSWORD);
+
+        Subscriber subscriber = subscriptionService.getSubscriber(callingNumber);
+        int numberOfSubsBefore = subscriber.getSubscriptions().size();
+
+        HttpPost httpPost2 = createSubscriptionHttpPost(callingNumber, "pack2");
+
+        SimpleHttpClient.execHttpRequest(httpPost2, HttpStatus.SC_OK, ADMIN_USERNAME, ADMIN_PASSWORD);
+
+        subscriber = subscriptionService.getSubscriber(callingNumber);
+        int numberOfSubsAfter = subscriber.getSubscriptions().size();
+
+        // Another subscription should be allowed because these are two different packs
+        assertTrue((numberOfSubsBefore + 1) == numberOfSubsAfter);
+    }
+
+    @Test
+    public void testCreateSubscriptionViaMcts() {
+        setupData();
+
+        Subscriber mctsSubscriber = new Subscriber(9999911122L);
+        mctsSubscriber.setDateOfBirth(LocalDate.now().minusDays(14));
+        subscriberDataService.create(mctsSubscriber);
+
+        subscriptionService.createSubscription(9999911122L, "10", "pack1", SubscriptionMode.MCTS_IMPORT);
+
+        mctsSubscriber = subscriberDataService.findByCallingNumber(9999911122L);
+        assertEquals(1, mctsSubscriber.getSubscriptions().size());
+    }
+
+    @Test
+    public void testCreateDuplicateChildSubscriptionViaMcts() {
+        setupData();
+
+        Subscriber mctsSubscriber = new Subscriber(9999911122L);
+        mctsSubscriber.setDateOfBirth(LocalDate.now().minusDays(14));
+        subscriberDataService.create(mctsSubscriber);
+
+        subscriptionService.createSubscription(9999911122L, "10", "pack1", SubscriptionMode.MCTS_IMPORT);
+
+        // attempt to create subscription to the same pack -- should fail
+        subscriptionService.createSubscription(9999911122L, "10", "pack1", SubscriptionMode.MCTS_IMPORT);
+
+        mctsSubscriber = subscriberDataService.findByCallingNumber(9999911122L);
+        assertEquals(1, mctsSubscriber.getSubscriptions().size());
```

Comment:
Should this assert also be earlier, just to make sure it is the second call that fails and not the first.